streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatter.java

```diff
@@ -340,32 +355,67 @@ public SegmentSearchResult find(final long timestamp, final boolean includeValue
+                                                                                : ByteBuffer.wrap(segmentValue).getLong(timestampSegmentIndex - (TIMESTAMP_SIZE + VALUE_SIZE));
+                final int currValueSize = ByteBuffer.wrap(segmentValue).getInt(timestampSegmentIndex + TIMESTAMP_SIZE);
                 if (currValueSize >= 0) {
                     final byte[] value = new byte[currValueSize];
-                    final int valueSegmentIndex = segmentValue.length - cumValueSize;
+                    final int valueSegmentIndex = getValueSegmentIndex(order, currentCumValueSize, currValueSize);
                     System.arraycopy(segmentValue, valueSegmentIndex, value, 0, currValueSize);
                     if (currTimestamp <= toTime && currNextTimestamp > fromTime) {
-                        segmentSearchResults.add(new SegmentSearchResult(currIndex, currTimestamp, currNextTimestamp, value));
+                        currentCumValueSize += currValueSize;
+                        currentDeserIndex++;
+                        return new SegmentSearchResult(currentDeserIndex - 1, currTimestamp, currNextTimestamp, value);
                     }
                 }
-
                 // prep for next iteration
-                currNextTimestamp = currTimestamp;
+                currentCumValueSize += Math.max(currValueSize, 0);
+                currentDeserIndex++;
+            }
+            // search in segment expected to find result but did not
+            return null;
+        }
+
+        private boolean hasStillRecord(final long currTimestamp, final long currNextTimestamp, final ResultOrder order) {
+            return order.equals(ResultOrder.ASCENDING) ? currNextTimestamp != nextTimestamp : currTimestamp != minTimestamp;
+        }
+
+        private int getValueSegmentIndex(final ResultOrder order, final int currentCumValueSize, final int currValueSize) {
+            return order.equals(ResultOrder.ASCENDING) ? valuesStartingIndex + currentCumValueSize
+                                                       : segmentValue.length - (currentCumValueSize + currValueSize);
+        }
+
+        private int getTimestampIndex(final ResultOrder order, final int currIndex) {
+            return order.equals(ResultOrder.ASCENDING) ? valuesStartingIndex - ((currIndex + 1) * (TIMESTAMP_SIZE + VALUE_SIZE))
+                                                       : 2 * TIMESTAMP_SIZE + currIndex * (TIMESTAMP_SIZE + VALUE_SIZE);
+        }
+
+        private void findValuesStartingIndex() {
+            long currTimestamp = -1;
+            int currIndex = 0;
+            int timestampSegmentIndex = 0;
+            while (currTimestamp != minTimestamp) {
+                timestampSegmentIndex = 2 * TIMESTAMP_SIZE + currIndex * (TIMESTAMP_SIZE + VALUE_SIZE);
+                currTimestamp = ByteBuffer.wrap(segmentValue).getLong(timestampSegmentIndex);
```

Comment:
Given that we parse all timestamps, we should buffer them (similar to what we already do with `unpackedReversedTimestampAndValueSizes`? Would we re-use `unpackedReversedTimestampAndValueSizes` and put `valueSize` as optional and populate later only on-demand?

Reply:
> Would we re-use `unpackedReversedTimestampAndValueSizes` and put `valueSize` as optional and populate later only on-demand?

re-using variables and data structures used by other methods makes the code very complicated and error-prone. Since others may do unwanted changes to them.

Reply:
> Given that we parse all timestamps, we should buffer them (similar to what we already do with `unpackedReversedTimestampAndValueSizes`? Would we re-use `unpackedReversedTimestampAndValueSizes` and put `valueSize` as optional and populate later only on-demand?

@mjsax I think it never happens that the same instance of `SegmentValue` is able to call methods of the `PartiallyDeserializedSegmentValue` class. If this assumption is true, isn't it better to move my `find()` to the `LogicalSegmentIterator` class? Having it in the other class (`PartiallyDeserializedSegmentValue`) is really confusing and not a good practice since all methods there are related to each other via the deserHelper variables and data structures. Then my `find()` seems orphan and irrelative.

Reply:
> Given that we parse all timestamps, we should buffer them (similar to what we already do with `unpackedReversedTimestampAndValueSizes`? Would we re-use `unpackedReversedTimestampAndValueSizes` and put `valueSize` as optional and populate later only on-demand?

Please have a look at my latest comment.